leapp/reporting/__init__.py

```diff
@@ -5,102 +5,256 @@
-    """
-    plaintext = fields.String()
-    """
-    Render report message as plaintext
-    """
+    report = fields.JSON()
 
 
-class Report(Model):
-    """
-    Framework model used for reporting and presentation (see "renderers" field) purposes. The report can also carry
-    a special meaning using "flags" field.
-    """
+def _add_to_dict(data, path, value, leaf_list=False):
+    for p in path[:-1]:
+        if p not in data:
+            data[p] = {}
+        data = data[p]
 
-    __non_inheritable__ = True
+    if leaf_list:
+        data[path[-1]] = data.get(path[-1], []) + [value]
+    else:
+        if path[-1] in data:
+            raise ValueError('Path {} is already taken'.format('.'.join(path)))
+        data[path[-1]] = value
 
-    topic = ReportTopic
 
-    severity = fields.StringEnum(choices=['low', 'medium', 'high'])
+class BasePrimitive(object):
     """
-    Severity of the report entry
+    Report primitive (field) base class
     """
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        self._value = value
 
-    title = fields.String()
-    """
-    Title of the report entry
-    """
+    @property
+    def is_list(self):
+        """
+        Return whether report field is supposed to be list or dict.
+        In case of a dict there can be only one value in the path
+        """
+        return False
 
-    detail = fields.JSON()
-    """
-    Detail of the report entry as JSON data
-    """
+    @property
+    def value(self):
+        """Return report field value"""
+        return self._value
 
-    renderers = fields.Model(Renderers)
-    """
-    :class:`Renderers` describe how to render this report entry
-    """
+    @property
+    def path(self):
+        """Return report field destination path (nested dict description)"""
+        raise NotImplementedError
 
-    audience = fields.List(fields.StringEnum(choices=['developer', 'sysadmin']))
-    """
-    Who is the main audience of this report entry
-    """
+    def apply(self, report):
+        """Add report entry to the report dict based on provided path"""
+        _add_to_dict(report, self.path, self.value, self.is_list)
 
-    flags = fields.List(fields.StringEnum(choices=['inhibitor']))
-    """
-    Flags can give a special meaning to each report entry, currently supported flags:
 
-        inhibitor - This report entry will inhibit the upgrade process
-    """
+class Title(BasePrimitive):
+    """Report title"""
+    @property
+    def path(self):
+        return ('title',)
+
+
+class Summary(BasePrimitive):
+    """Report summary"""
+    @property
+    def path(self):
+        return ('summary',)
+
+
+class Severity(BasePrimitive):
+    """Report severity"""
+
+    INFO = 'info'
+    LOW = 'low'
+    MEDIUM = 'medium'
+    HIGH = 'high'
+
+    def __init__(self, value=None):
+        if not isinstance(value, str):
```

Comment:
Wondering if this check is necessary at all - no value other that one of 4 strings in severities will be allowed anyway. Do we really need such strict type checking (raising TypeError in one case and ValueError in another), won't check at L93 be enough?

Reply:
removed